lib/SILOptimizer/Mandatory/ClosureLifetimeFixup.cpp

```diff
@@ -712,7 +714,186 @@ static SILValue tryRewriteToPartialApplyStack(
+      if (initialization) {
+        LLVM_DEBUG(llvm::dbgs() << "-- had non-initialization, non-partial-apply use\n";
+                   use->getUser()->print(llvm::dbgs());
+                   llvm::dbgs() << "\n");
+                   
+        initialization = nullptr;
+        break;
+      }
+      if (auto possibleInit = dyn_cast<CopyAddrInst>(use->getUser())) {
+        // Should copy the source and initialize the destination.
+        if (possibleInit->isTakeOfSrc()
+            || !possibleInit->isInitializationOfDest()) {
+          LLVM_DEBUG(llvm::dbgs() << "-- had non-initialization, non-partial-apply use\n";
+                     use->getUser()->print(llvm::dbgs());
+                     llvm::dbgs() << "\n");
+
+          break;
+        }
+        // This is the initialization if there are no other uses.
+        initialization = possibleInit;
+        continue;
+      }
+      LLVM_DEBUG(llvm::dbgs() << "-- unrecognized use\n");
+      break;
+    }
+    if (!initialization) {
+      LLVM_DEBUG(llvm::dbgs() << "-- failed to find single initializing use\n");
+      continue;
+    }
+    
+    // The source should have no writes in the duration of the partial_apply's
+    // liveness.
+    auto orig = initialization->getSrc();
+    LLVM_DEBUG(llvm::dbgs() << "++ found original:\n";
+               orig->print(llvm::dbgs());
+               llvm::dbgs() << "\n");
+               
+    bool origIsUnusedDuringClosureLifetime = true;
+
+    class OrigUnusedDuringClosureLifetimeWalker
+      : public TransitiveAddressWalker
+    {
+      SSAPrunedLiveness &closureLiveness;
+      bool &origIsUnusedDuringClosureLifetime;
+    public:
+      OrigUnusedDuringClosureLifetimeWalker(SSAPrunedLiveness &closureLiveness,
+                                        bool &origIsUnusedDuringClosureLifetime)
+        : closureLiveness(closureLiveness),
+          origIsUnusedDuringClosureLifetime(origIsUnusedDuringClosureLifetime)
+      {}
+    
+      virtual bool visitUse(Operand *origUse) override {
+        LLVM_DEBUG(llvm::dbgs() << "looking at use\n";
+                   origUse->getUser()->printInContext(llvm::dbgs());
+                   llvm::dbgs() << "\n");
+        
+        // If the user doesn't write to memory, then it's harmless.
+        if (!origUse->getUser()->mayWriteToMemory()) {
+          return true;
+        }
+        if (closureLiveness.isWithinBoundary(origUse->getUser())) {
+          origIsUnusedDuringClosureLifetime = false;
+          LLVM_DEBUG(llvm::dbgs() << "-- original has other possibly writing use during closure lifetime\n";
+                     origUse->getUser()->print(llvm::dbgs());
+                     llvm::dbgs() << "\n");
+          return false;
+        }
+        return true;
+      }
+    };
+    
+    OrigUnusedDuringClosureLifetimeWalker origUseWalker(closureLiveness,
+                                              origIsUnusedDuringClosureLifetime);
+    std::move(origUseWalker).walk(orig);
```

Comment:
You may want for future proofing to add the check that we do not return AddressUseKind::Unknown. This is propagated back when you return false from the walk function. That being said, I do see that in the current iteration of the code origIsUnusedDuringClosureLifetime has the same effect. Your choice on what you want to do with this.

Reply:
If the result is anything other than `AddressUseKind::NonEscaping`, then the walker did not find all address uses. The problem is that other closure captures will report an escaping use even if the address only escapes immutably.

If this is too conservative, then my previous suggestion was to replace this code in the walker:
```
// TODO: Partial apply should be NonEscaping, but then we need to consider
    // the apply to be a use point.
    if (isa<PartialApplyInst>(user) || isa<AddressToPointerInst>(user)) {
      meet(AddressUseKind::PointerEscape);
      callVisitUse(op);
      continue;
    }
```
With a `visitCapture` override that returns `AddressUseKind`, which `OrigUnusedDuringClosureLifetimeWalker` could implement to return `NonEscaping` for immutable captures and `PointerEscape` for mutable captures.

Reply:
I tried to explain this better here: rdar://111160731 ([move-only] Handle pointer escapes in OrigUnusedDuringClosureLifetimeWalker)